pyomo/gdp/plugins/binary_multiplication.py

```diff
@@ -0,0 +1,173 @@
+        for disjunct in obj.disjuncts:
+            or_expr += disjunct.binary_indicator_var
+            self._transform_disjunct(disjunct, transBlock)
+
+        rhs = 1 if parent_disjunct is None else parent_disjunct.binary_indicator_var
+        if obj.xor:
+            xorConstraint[index] = or_expr == rhs
+        else:
+            xorConstraint[index] = or_expr >= rhs
+        # Mark the DisjunctionData as transformed by mapping it to its XOR
+        # constraint.
+        obj._algebraic_constraint = weakref_ref(xorConstraint[index])
+
+        # and deactivate for the writers
+        obj.deactivate()
+
+    def _transform_disjunct(self, obj, transBlock):
+        # We're not using the preprocessed list here, so this could be
+        # inactive. We've already done the error checking in preprocessing, so
+        # we just skip it here.
+        if not obj.active:
+            return
+
+        relaxationBlock = self._get_disjunct_transformation_block(obj, transBlock)
+
+        # Transform each component within this disjunct
+        self._transform_block_components(obj, obj)
+
+        # deactivate disjunct to keep the writers happy
+        obj._deactivate_without_fixing_indicator()
+
+    def _transform_constraint(self, obj, disjunct):
+        # add constraint to the transformation block, we'll transform it there.
+        transBlock = disjunct._transformation_block()
+        constraintMap = transBlock._constraintMap
+
+        disjunctionRelaxationBlock = transBlock.parent_block()
+
+        # We will make indexes from ({obj.local_name} x obj.index_set() x ['lb',
+        # 'ub']), but don't bother construct that set here, as taking Cartesian
+        # products is kind of expensive (and redundant since we have the
+        # original model)
+        newConstraint = transBlock.transformedConstraints
+
+        for i in sorted(obj.keys()):
+            c = obj[i]
+            if not c.active:
+                continue
+
+            self._add_constraint_expressions(
+                c, i, disjunct.binary_indicator_var, newConstraint, constraintMap
+            )
+
+            # deactivate because we relaxed
+            c.deactivate()
+
+    def _add_constraint_expressions(
+        self, c, i, indicator_var, newConstraint, constraintMap
+    ):
+        # Since we are both combining components from multiple blocks and using
+        # local names, we need to make sure that the first index for
+        # transformedConstraints is guaranteed to be unique. We just grab the
+        # current length of the list here since that will be monotonically
+        # increasing and hence unique. We'll append it to the
+        # slightly-more-human-readable constraint name for something familiar
+        # but unique. (Note that we really could do this outside of the loop
+        # over the constraint indices, but I don't think it matters a lot.)
+        unique = len(newConstraint)
+        name = c.local_name + "_%s" % unique
+
```

Comment:
Can we make this a little simpler by starting with:
```python
constraintMap['transformedConstraints'][c] = transformed = []
```
Then everything else can unconditionally append to `transformed`?